[LTC] Add support for non-structured in-place operator variants

alanwaketan opened the pull request:
Summary:
This commit teached the LTC code-gen engine about in-place operator variants and generated zero_ for validation.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestZeroInPlace

Fixes #65576.

Commit:
[LTC] Add support for non-structured in-place operator variants

Summary:
This commit teached the LTC code-gen engine about in-place operator variants and generated zero_ for validation.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestZeroInPlace

Commit:
[LTC] Code-gen relu/relu_

Summary:
This commit code-generated relu and relu_ and teached the code-gen engine
to know how to avoid re-generating IR classes and shape inferences for
in-place variants.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc --gtest_filter=AtenLtcTsTensorTest.TestRelu*

Commit:
[LTC] Refactor how shape functions are declared and called

Summary:
This commit added a helper class for shape functions such that it's
easier to maintain in the future.

Test Plan:
lazy_tensor_core/test/cpp/build/test_ptltc

wconstab reviewed (approved):
Looks good to me.  I like the new choice of new 'ComputeShapeSignature' to clean things up a bit.

Requested one comment added.

Commit:
[LTC] Refactor gen_lazy_nativefunc_definition to a functor

Summary:
This commit refactorred gen_lazy_nativefunc_definition to GenLazyNativeFuncDefinition
such that it can utilize the native function context manager while accepting
additional parameters.

Test Plan:
azy_tensor_core/test/cpp/build/test_ptltc
